================================================================
Index: Expat/Expat.xs
===================================================================
RCS file: /home/ccc/CVS/Parser/Expat/Expat.xs,v
retrieving revision 1.23
diff -c -r1.23 Expat.xs
*** Expat.xs	1999/07/25 23:07:51	1.23
--- Expat.xs	1999/08/26 19:05:57
***************
*** 33,38 ****
--- 33,41 ----
  
  #define DTB_GROW 4096
  
+ /* Macro to update handler fields. Used in the various handler setting
+    XSUBS */
+ 
  #define XMLP_UPD(fld) \
    RETVAL = cbv->fld ? newSVsv(cbv->fld) : &PL_sv_undef;\
    if (cbv->fld) {\
***************
*** 42,47 ****
--- 45,57 ----
    else\
      cbv->fld = newSVsv(fld)
  
+ /* Macro to push old handler value onto return stack. This is done here
+    to get around a bug in 5.004 sv_2mortal function. */
+ 
+ #define PUSHRET \
+   ST(0) = RETVAL;\
+   if (RETVAL != &PL_sv_undef && SvREFCNT(RETVAL)) sv_2mortal(RETVAL)
+ 
  /* These are flags set in the dflags field. They indicate whether the
     corresponding handler is set. All these handlers actually use expat's
     default handler. By setting and unsetting these in dflags, we can
***************
*** 2012,2020 ****
  	{
  	  CallbackVector * cbv = (CallbackVector*) XML_GetUserData(parser);
  	  XMLP_UPD(start_sv);
  	}
-     OUTPUT:
-         RETVAL
  
  SV *
  XML_SetEndElementHandler(parser, end_sv)
--- 2022,2029 ----
  	{
  	  CallbackVector * cbv = (CallbackVector*) XML_GetUserData(parser);
  	  XMLP_UPD(start_sv);
+ 	  PUSHRET;
  	}
  
  SV *
  XML_SetEndElementHandler(parser, end_sv)
***************
*** 2024,2032 ****
  	{
  	  CallbackVector *cbv = (CallbackVector*) XML_GetUserData(parser);
  	  XMLP_UPD(end_sv);
  	}
-     OUTPUT:
-         RETVAL
  
  SV *
  XML_SetCharacterDataHandler(parser, char_sv)
--- 2033,2040 ----
  	{
  	  CallbackVector *cbv = (CallbackVector*) XML_GetUserData(parser);
  	  XMLP_UPD(end_sv);
+ 	  PUSHRET;
  	}
  
  SV *
  XML_SetCharacterDataHandler(parser, char_sv)
***************
*** 2042,2050 ****
  	    charhndl = characterData;
  
  	  XML_SetCharacterDataHandler(parser, charhndl);
  	}
-     OUTPUT:
-         RETVAL
  
  SV *
  XML_SetProcessingInstructionHandler(parser, proc_sv)
--- 2050,2057 ----
  	    charhndl = characterData;
  
  	  XML_SetCharacterDataHandler(parser, charhndl);
+ 	  PUSHRET;
  	}
  
  SV *
  XML_SetProcessingInstructionHandler(parser, proc_sv)
***************
*** 2061,2069 ****
  	    prochndl = processingInstruction;
  
  	  XML_SetProcessingInstructionHandler(parser, prochndl);
  	}
-     OUTPUT:
-         RETVAL
  
  SV *
  XML_SetCommentHandler(parser, cmnt_sv)
--- 2068,2075 ----
  	    prochndl = processingInstruction;
  
  	  XML_SetProcessingInstructionHandler(parser, prochndl);
+ 	  PUSHRET;
  	}
  
  SV *
  XML_SetCommentHandler(parser, cmnt_sv)
***************
*** 2079,2087 ****
  	    cmnthndl = commenthandle;
  
  	  XML_SetCommentHandler(parser, cmnthndl);
  	}
-     OUTPUT:
-         RETVAL
  
  SV *
  XML_SetDefaultHandler(parser, dflt_sv)
--- 2085,2092 ----
  	    cmnthndl = commenthandle;
  
  	  XML_SetCommentHandler(parser, cmnthndl);
+ 	  PUSHRET;
  	}
  
  SV *
  XML_SetDefaultHandler(parser, dflt_sv)
***************
*** 2097,2105 ****
  	    set = 1;
  
  	  check_and_set_default_handler(parser, cbv, set, INST_DFL);
  	}
-     OUTPUT:
-         RETVAL
  
  SV *
  XML_SetUnparsedEntityDeclHandler(parser, unprsd_sv)
--- 2102,2109 ----
  	    set = 1;
  
  	  check_and_set_default_handler(parser, cbv, set, INST_DFL);
+ 	  PUSHRET;
  	}
  
  SV *
  XML_SetUnparsedEntityDeclHandler(parser, unprsd_sv)
***************
*** 2116,2124 ****
  	    unprsdhndl = unparsedEntityDecl;
  
  	  XML_SetUnparsedEntityDeclHandler(parser, unprsdhndl);
  	}
-     OUTPUT:
-         RETVAL
  
  SV *
  XML_SetNotationDeclHandler(parser, notation_sv)
--- 2120,2127 ----
  	    unprsdhndl = unparsedEntityDecl;
  
  	  XML_SetUnparsedEntityDeclHandler(parser, unprsdhndl);
+ 	  PUSHRET;
  	}
  
  SV *
  XML_SetNotationDeclHandler(parser, notation_sv)
***************
*** 2134,2142 ****
  	    nothndlr = notationDecl;
  
  	  XML_SetNotationDeclHandler(parser, nothndlr);
  	}
-     OUTPUT:
-         RETVAL
  
  SV *
  XML_SetExternalEntityRefHandler(parser, extent_sv)
--- 2137,2144 ----
  	    nothndlr = notationDecl;
  
  	  XML_SetNotationDeclHandler(parser, nothndlr);
+ 	  PUSHRET;
  	}
  
  SV *
  XML_SetExternalEntityRefHandler(parser, extent_sv)
***************
*** 2153,2161 ****
  	    exthndlr = externalEntityRef;
  
  	  XML_SetExternalEntityRefHandler(parser, exthndlr);
  	}
-     OUTPUT:
-         RETVAL
  	   
  SV *
  XML_SetEntityDeclHandler(parser, entdcl_sv)
--- 2155,2162 ----
  	    exthndlr = externalEntityRef;
  
  	  XML_SetExternalEntityRefHandler(parser, exthndlr);
+ 	  PUSHRET;
  	}
  	   
  SV *
  XML_SetEntityDeclHandler(parser, entdcl_sv)
***************
*** 2171,2179 ****
  	    set = 1;
  
  	  check_and_set_default_handler(parser, cbv, set, INST_ENT);
  	}
-     OUTPUT:
-         RETVAL
  
  SV *
  XML_SetElementDeclHandler(parser, eledcl_sv)
--- 2172,2179 ----
  	    set = 1;
  
  	  check_and_set_default_handler(parser, cbv, set, INST_ENT);
+ 	  PUSHRET;
  	}
  
  SV *
  XML_SetElementDeclHandler(parser, eledcl_sv)
***************
*** 2189,2197 ****
  	    set = 1;
  
  	  check_and_set_default_handler(parser, cbv, set, INST_ELE);
  	}
-     OUTPUT:
-         RETVAL
  
  SV *
  XML_SetAttListDeclHandler(parser, attdcl_sv)
--- 2189,2196 ----
  	    set = 1;
  
  	  check_and_set_default_handler(parser, cbv, set, INST_ELE);
+ 	  PUSHRET;
  	}
  
  SV *
  XML_SetAttListDeclHandler(parser, attdcl_sv)
***************
*** 2207,2215 ****
  	    set = 1;
  
  	  check_and_set_default_handler(parser, cbv, set, INST_ATT);
  	}
-     OUTPUT:
-         RETVAL
  
  SV *
  XML_SetDoctypeHandler(parser, doctyp_sv)
--- 2206,2213 ----
  	    set = 1;
  
  	  check_and_set_default_handler(parser, cbv, set, INST_ATT);
+ 	  PUSHRET;
  	}
  
  SV *
  XML_SetDoctypeHandler(parser, doctyp_sv)
***************
*** 2225,2233 ****
  	    set = 1;
  
  	  check_and_set_default_handler(parser, cbv, set, INST_DOC);
  	}
-     OUTPUT:
-         RETVAL
  
  SV *
  XML_SetXMLDeclHandler(parser, xmldec_sv)
--- 2223,2230 ----
  	    set = 1;
  
  	  check_and_set_default_handler(parser, cbv, set, INST_DOC);
+ 	  PUSHRET;
  	}
  
  SV *
  XML_SetXMLDeclHandler(parser, xmldec_sv)
***************
*** 2243,2251 ****
  	    set = 1;
  
  	  check_and_set_default_handler(parser, cbv, set, INST_XML);
  	}
-     OUTPUT:
-         RETVAL
  
  int
  XML_SetBase(parser, base)
--- 2240,2247 ----
  	    set = 1;
  
  	  check_and_set_default_handler(parser, cbv, set, INST_XML);
+ 	  PUSHRET;
  	}
  
  int
  XML_SetBase(parser, base)
***************
*** 2592,2600 ****
  	    ecdhndl = endCdata;
  
  	  XML_SetCdataSectionHandler(parser, scdhndl, ecdhndl);
  	}
-     OUTPUT:
-         RETVAL
  
  SV *
  XML_SetEndCdataHandler(parser, endcd_sv)
--- 2588,2595 ----
  	    ecdhndl = endCdata;
  
  	  XML_SetCdataSectionHandler(parser, scdhndl, ecdhndl);
+ 	  PUSHRET;
  	}
  
  SV *
  XML_SetEndCdataHandler(parser, endcd_sv)
***************
*** 2616,2624 ****
  	    scdhndl = startCdata;
  
  	  XML_SetCdataSectionHandler(parser, scdhndl, ecdhndl);
  	}
-     OUTPUT:
-         RETVAL
  
  void
  XML_UnsetAllHandlers(parser)
--- 2611,2618 ----
  	    scdhndl = startCdata;
  
  	  XML_SetCdataSectionHandler(parser, scdhndl, ecdhndl);
+ 	  PUSHRET;
  	}
  
  void
  XML_UnsetAllHandlers(parser)

