*** xmlparse.c.orig	Thu Sep 24 11:37:24 1998
--- xmlparse.c	Fri Sep 25 17:54:07 1998
***************
*** 774,779 ****
--- 774,781 ----
  void XML_DefaultCurrent(XML_Parser parser)
  {
    if (defaultHandler) {
+     const char *eventPHold = eventPtr;
+     const char *eventEndPHold = eventEndPtr;
      if (openInternalEntities)
        reportDefault(parser,
  	            ns ? XmlGetInternalEncodingNS() : XmlGetInternalEncoding(),
***************
*** 781,786 ****
--- 783,791 ----
  		    openInternalEntities->internalEventEndPtr);
      else
        reportDefault(parser, encoding, eventPtr, eventEndPtr);
+ 
+     eventPtr = eventPHold;
+     eventEndPtr = eventEndPHold;
    }
  }
  
